add BLE to bootloader

TychoVrahe opened the pull request:
This PR adds support for BLE in bootloader.

Event polling in bootloader is refactored to use the new sysevents mechanism.

Some rust layouts that need event processing while USB/BLE communication is running are now allocated on stack in C. 

Cancel button is added to ''waiting for host' screen in all bootloader UIs.

BLE support is implemented in Bolt bootloader UI.

Further work needed in successive PRs:
- solve stopping/enabling communication issues when switching screens/flows (busy response, disconnect USB, ...)
- add power off button to menu ( https://github.com/trezor/trezor-firmware/pull/4878)

<!--
For core developers:
- Assign yourself to the PR.
- Set the priority to match the original issue.
- Add the PR to the current sprint.
- If it's a draft PR, mark it as "In Progress."
- If it's a final PR, mark it as "Needs Review."

For external contributors:
- Please open an issue before submitting a PR so we can discuss whether we want to proceed with it.
-->

TychoVrahe commented on core/embed/projects/bootloader/workflow/wf_ble_pairing_request.c:
```diff
@@ -0,0 +1,131 @@
+#include "workflow.h"
+
+static bool encode_pairing_code(uint32_t code, uint8_t *outbuf) {
+  if (code < 0 || code > 999999) {
```
fixed https://github.com/trezor/trezor-firmware/commit/a0b8689072c6d5d7f4068695b30339c8d5392af4

TychoVrahe commented on core/embed/sys/task/sysevent.c:
```diff
@@ -221,7 +221,7 @@ void sysevents_poll(const sysevents_t *awaited, sysevents_t *signalled,
                    (poller->signalled->write_ready != 0);
       if (ready || timed_out) {
-        systask_t *task = poller->task;
+        systask_t *task = poller->task;  // USB interface
```
removed, not sure how it got there: https://github.com/trezor/trezor-firmware/commit/e14b492e596ba636279bdde302e5f30e28012685

TychoVrahe commented on core/SConscript.bootloader_emu:
```diff
@@ -184,6 +183,7 @@ ALLPATHS = ['embed/rust',
            'embed/gfx/inc',
            'embed/io/display/inc',
            'embed/io/usb/inc',
+           'embed/io/poll/inc',
```
removed https://github.com/trezor/trezor-firmware/commit/bd584edbe5311d3bc52ddef2dd98c4df76fb1c27

TychoVrahe commented on core/Makefile:
```diff
@@ -30,10 +30,10 @@ PRODUCTION ?= 0
+BOOTLOADER_DEVEL ?= 1
 DISABLE_OPTIGA ?= 0
 DISABLE_TROPIC ?= 1
-TREZOR_MODEL ?= T2T1
```
sorry, these were already removed during the last force-push

TychoVrahe commented on core/embed/projects/bootloader/wire/wire_iface_ble.c:
```diff
@@ -0,0 +1,173 @@
+    };
+    ble_issue_command(&cmd_disconnect);
+    systick_delay_ms(20);
+    ble_get_state(&state);
```
added some checks https://github.com/trezor/trezor-firmware/commit/6241c0b9864de1e9f2ad73e965cdc3ebf46db3c0

TychoVrahe commented on core/embed/projects/bootloader/wire/wire_iface_ble.c:
```diff
@@ -0,0 +1,173 @@
+                       .name = "Trezor Bootloader",
+                       .static_mac = false,
+                   }},
+      .data_len = 0,
```
removed the zero, it should work fine without it: https://github.com/trezor/trezor-firmware/commit/6241c0b9864de1e9f2ad73e965cdc3ebf46db3c0

but i am thinking of redoing the issue_command (in next PR) and make the commands separate functions with clearer interface. At the cost of several more syscalls, but the usage will be much prettier. What do you think?

TychoVrahe commented on core/embed/projects/bootloader/wire/wire_iface_ble.c:
```diff
@@ -0,0 +1,173 @@
+  iface->read = &ble_read_;
+  iface->error = &ble_error;
+
+  ble_start();
```
hmm i don't really want to deinit the whole BLE driver, as i still want the state monitored in background. Maybe we should `stop` USB instead of `deinit` too. But as i mentioned in PR description, there is still some work needed on deinitialization of USB/BLE communication when switching screens and this is closely related, so i'd like to focus on that in next PR.

TychoVrahe commented on core/embed/projects/bootloader/wire/wire_iface_ble.c:
```diff
@@ -0,0 +1,208 @@
+  }
+
+  ble_command_t cmd = {
+      .cmd_type = BLE_PAIRING_MODE,
```
i agree the names are not ideal - as discussed in https://github.com/trezor/trezor-firmware/pull/4833#discussion_r2039598588 and in person, i will replace the commands with dedicated functions in next PR, so i will handle renaming while doing that.

TychoVrahe commented on core/embed/projects/bootloader/workflow/wf_ble_pairing_request.c:
```diff
@@ -0,0 +1,119 @@
+  if (code > 999999) {
+    return false;
+  }
+  for (size_t i = 0; i < BLE_PAIRING_CODE_LEN; i++) {
```
i kinda don't want to pull these functions in the build just for this, also i don't use null terminated string

TychoVrahe commented:
@cepetr in this commit https://github.com/trezor/trezor-firmware/pull/4833/commits/d8e3fb8b76439e0a2bb9e3e85fa05a92d84de266, i introduce a new type for the layout buffer, so it simplifies some calls a bit

cepetr commented on core/embed/projects/bootloader/workflow/wf_host_control.c:
```diff
@@ -109,6 +152,9 @@ workflow_result_t workflow_host_control(const vendor_header *const vhdr,
-  usb_iface_deinit(&usb_iface);
+  usb_iface_deinit();
+#ifdef USE_BLE
+  ble_iface_init();
```
`ble_iface_deinit()`?

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/commit/cce56b70f4fb93a3c957801ed7f5952e2382f1e3

TychoVrahe commented on core/embed/projects/bootloader/wire/wire_iface_ble.c:
```diff
@@ -0,0 +1,208 @@
+      return false;
+    }
+    if (!is_connected()) {
+      return false;
```
fixed https://github.com/trezor/trezor-firmware/commit/cce56b70f4fb93a3c957801ed7f5952e2382f1e3

TychoVrahe commented on core/embed/rust/src/ui/layout_caesar/bootloader/mod.rs:
```diff
@@ -273,9 +307,12 @@ impl BootloaderUI for UICaesar {
         let mut frame = Connect::new("Waiting for host...", fonts::FONT_NORMAL, BLD_FG, BLD_BG);
-        show(&mut frame, false);
+
+        show(&mut frame, true);
```
removed fading: https://github.com/trezor/trezor-firmware/commit/cce56b70f4fb93a3c957801ed7f5952e2382f1e3

TychoVrahe commented on core/embed/rust/src/ui/layout_bolt/bootloader/welcome.rs:
```diff
@@ -1,45 +1,74 @@
+        if let Event::Button(_) = _event {
+            return Some(WelcomeMsg::PairingMode);
+        }
+        #[cfg(feature = "powerctl")]
```
unified: https://github.com/trezor/trezor-firmware/commit/cce56b70f4fb93a3c957801ed7f5952e2382f1e3

TychoVrahe commented on core/embed/projects/bootloader/main.c:
```diff
@@ -390,6 +401,11 @@ int bootloader_main(void) {
       result = workflow_empty_device();
     }
 
+    usb_iface_deinit();
```
agree, see comment above, i think it falls under the same problem: https://github.com/trezor/trezor-firmware/pull/4833#discussion_r2039606064

TychoVrahe commented on core/embed/rust/rust_ui_bootloader.h:
```diff
@@ -1,29 +1,91 @@
+typedef enum {
+  CANCEL = 1,
+  CONFIRM = 2,
+} confirm_result_t;
```
agree. the style used in this file would suggest naming `CONFIRM_CANCEL` and `CONFIRM_CONFIRM`, which i thought was very ugly. 

anyway, lets leave this for further work with hope that we will be able to get rid of these repeated definitions

cepetr reviewed (approved):
tested on t2t1, t3b1, t3t1, partially on t3w1

matejcik reviewed (commented):
the unsafety in the Rust-C buffer interface is rather awkwardly encapsulated, and `SAFETY` comments are missing in most places.

rather than commenting item by item, I designed a somewhat different unsafety encapsulation, please review (probably @cepetr) and cherry pick 1c07cdc7841993498b964b184b838407da7787fc

matejcik commented:
one more fixup in https://github.com/trezor/trezor-firmware/compare/tychovrahe/T3W1/ble_bootloader_pr...matejcik/ble_bootloader_fixup

TychoVrahe commented:
Cherry picked @matejcik commits, tested on device and made few fixes in https://github.com/trezor/trezor-firmware/commit/2d446c75fabf36a857e9c2c96d59cf39cf3569d8

Mainly: added missing show() calls, fixed writing into layout memory, fixed some clippy complains about missing mut.

matejcik reviewed (approved):
ACK on https://github.com/trezor/trezor-firmware/commit/2d446c75fabf36a857e9c2c96d59cf39cf3569d8

cepetr commented on core/embed/rust/src/trezorhal/layout_buf.rs:
```diff
@@ -0,0 +1,69 @@
+    /// of the `LayoutBuffer`.
+    pub unsafe fn new(buffer: *mut c_layout_t) -> Self {
+        let layout = unwrap!(NonNull::new(buffer));
+        // SAFETY: the layout is non-null and properly aligned
```
IMHO, no one guarantees that `c_layout_t` is properly aligned - since it’s allocated on the stack, it’s currently aligned to 4 bytes. That might be sufficient for all layouts and could always work, but wouldn’t it be safer to split the array using `align_to_mut()` here and use the properly aligned pointer?

matejcik replied:
you're right, I misread NonNull docs and thought that it guarantees alignment which it doesn't.

i believe it doesn't matter with the types that we're using, because `c_layout_t` is just an array of u8 which is always aligned ... and the alignment requirement for `T` is explicitly checked a couple lines down

i'm wondering if we need the whole song-and-dance about `c_layout_t` struct? but that's probably the easiest way to ensure the buffer size across the C-Rust boundary, correct? but if we could just pass a buffer + size from C to Rust, we wouldn't need to worry about changes to `c_layout_t` struct

matejcik replied:
at minimum we can fix the comment
```suggestion
        // SAFETY: c_layout_t is just an array of u8 which is always aligned
```

cepetr replied:
I think the problem is that `c_layout_t` is defined as a struct containing a `uint8_t` array - so the pointer to this array is not necessarily aligned. References to this array might be typically aligned to 4 bytes, but I would not rely on it.

If the type `T` requires a higher alignment (e.g., 8 bytes), the code may panic .

TychoVrahe replied:
how about this? https://github.com/trezor/trezor-firmware/commit/4dbfbac0fba7be34963db43f844c1618525eb4d5

enforcing the alignment via proposed align_to_mut 

(in successive work, i used c_layout_t allocated statically, which was then not aligned to 4 bytes and this paniced as prediced, i went for enforcing alignment via gcc attribute but this seems more robust.

cepetr replied:
Looks good. Maybe we can try to get rid of redundant alignment 

and instead of storing of `layout: NonNull<c_layout_t>,`
store the aligned pointer as `layout: NonNull<MaybeUninit<T>>, 

so that `get_maybe_uninit()` would be simpler

TychoVrahe replied:
one more variant: https://github.com/trezor/trezor-firmware/commit/e74a04c696710d0663df97a8abe70b31cba23cae

IIUC we still need the marker as NonNull is essentially a pointer and we expect ownership of underlying data

cepetr reviewed (approved):
I went through the latest changes - lgtm. I just have a small note regarding pointer alignment in LayoutBuffer::new.
